File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;
        int N = Integer.parseInt(br.readLine());
        ArrayList<Integer>[] list = new ArrayList[N + 1];
        int[] parent = new int[N + 1];
        for(int i = 1; i <= N; i++) {
            list[i] = new ArrayList<>();
        }
        for(int i = 0; i < N - 1; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            list[a].add(b);
            list[b].add(a);
        }
        Queue<Integer> que = new LinkedList<>();
        parent[1] = 1;
        que.add(1);
        while(!que.isEmpty()) {
            int node = que.poll();
            for(int num : list[node]) {
                if(parent[num] != 0) continue;
                parent[num] = node;
                que.add(num);
            }
        }
        for(int i = 2; i <= N; i++) {
            sb.append(parent[i]).append("\n");
        }
        System.out.print(sb);
    }
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.js
class Queue {
  constructor() {
    this.arr = [];
    this.head = 0;
  }

  add(num) {
    this.arr.push(num);
  }

  poll() {
    return this.arr[this.head++];
  }

  isEmpty() {
    return this.arr.length === this.head;
  }
}

let [N, ...input] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n').map(line => line.trim());
N = Number(N);
list = Array.from({length : N + 1}, () => []);
input = input.map(line => line.split(' ').map(Number));
for(let i = 0; i < N - 1; i++) {
  list[input[i][0]].push(input[i][1]);
  list[input[i][1]].push(input[i][0]);
}
const que = new Queue();
const parent = new Array(N + 1).fill(0);
que.add(1);
parent[1] = 1;
while(!que.isEmpty()) {
  const node = que.poll();
  for(let num of list[node]) {
    if(parent[num] != 0) continue;
    parent[num] = node;
    que.add(num);
  }
}
console.log(parent.slice(2).join('\n'));
